test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java

```diff
@@ -356,6 +359,80 @@ public void testGetDocContent() throws Exception {
+    DocId docId = newDocId(new Id(id));
+    MockObjectStore os = getObjectStore();
+
+    mockDocument(os, id, DOCUMENT_TIMESTAMP, true,
+        1000d, "text/plain",
+        TestObjectFactory.getPermissions(
+            PermissionSource.SOURCE_DIRECT,
+            PermissionSource.SOURCE_TEMPLATE,
+            PermissionSource.SOURCE_PARENT),
+        new ActiveMarkingListMock(
+            mockActiveMarking(markingId1,
+                TestObjectFactory.getPermissions(PermissionSource.MARKING)),
+            mockActiveMarking(markingId2,
+                TestObjectFactory.getPermissions(PermissionSource.MARKING))));
+
+    DocumentTraverser traverser = new DocumentTraverser(options);
+    RecordingResponse response = new RecordingResponse();
+    traverser.getDocContent(new Id(id), null, response);
+
+    assertEquals(
+        ImmutableSet.of(PropertyNames.ID, PropertyNames.DATE_LAST_MODIFIED,
+            PropertyNames.CONTENT_SIZE, PropertyNames.MIME_TYPE),
+        response.getMetadata().getKeys());
+
+    assertEquals("text/plain", response.getContentType());
+    byte[] actualContent =
+        ((ByteArrayOutputStream) response.getOutputStream()).toByteArray();
+    assertEquals("sample content", new String(actualContent, UTF_8));
+
+    Acl acl = response.getAcl();
+    assertFalse(acl.getPermitUsers().toString(),
+        acl.getPermitUsers().isEmpty());
+    assertEquals(docId, acl.getInheritFrom());
+    assertEquals("TMPL", acl.getInheritFromFragment());
+
+    acl = response.getNamedResources().get("TMPL");
+    assertFalse(acl.getPermitUsers().toString(),
+        acl.getPermitUsers().isEmpty());
+    assertEquals(InheritanceType.CHILD_OVERRIDES, acl.getInheritanceType());
+    assertEquals(docId, acl.getInheritFrom());
+    assertEquals("FLDR", acl.getInheritFromFragment());
+
+    acl = response.getNamedResources().get("FLDR");
+    assertFalse(acl.getPermitUsers().toString(),
+        acl.getPermitUsers().isEmpty());
+    assertEquals(InheritanceType.CHILD_OVERRIDES, acl.getInheritanceType());
+    assertEquals(docId, acl.getInheritFrom());
+    assertEquals(markingFrag(markingId2), acl.getInheritFromFragment());
+
+    acl = response.getNamedResources().get(markingFrag(markingId2));
+    assertFalse(acl.getPermitUsers().toString(),
+        acl.getPermitUsers().isEmpty());
+    assertEquals(InheritanceType.AND_BOTH_PERMIT, acl.getInheritanceType());
+    assertEquals(docId, acl.getInheritFrom());
+    assertEquals(markingFrag(markingId1), acl.getInheritFromFragment());
+
+    acl = response.getNamedResources().get(markingFrag(markingId1));
+    assertFalse(acl.getPermitUsers().toString(),
+        acl.getPermitUsers().isEmpty());
+    assertEquals(InheritanceType.AND_BOTH_PERMIT, acl.getInheritanceType());
+    assertEquals(null, acl.getInheritFrom());
+    assertEquals(null, acl.getInheritFromFragment());
+  }
+
+  private String markingFrag(String id) {
+    return "MARK" + id.replace("{", "%7B").replace("}", "%7D");
```

Comment:
We have this code twice in DocumentTraverser now, too, so maybe add a helper method like 

String percentEscape(Id)